tests/test_mem_eff_attention.py

```diff
@@ -48,49 +134,305 @@ def test_key_query_all_ones(device, q_len, kv_len, batch_size, k_len):
+    ],
+)
+def test_logsumexp(
+    device,
+    q_len,
+    kv_len,
+    batch_size,
+    k_len,
+    dtype,
+    op: xformers.ops.MemoryEfficientAttentionOp,
+):
+    scale = 3
+    query = torch.randn((batch_size, q_len, k_len), device=device, dtype=dtype) * scale
+    key = torch.randn((batch_size, kv_len, k_len), device=device, dtype=dtype) * scale
+    value = torch.randn((batch_size, kv_len, k_len), device=device, dtype=dtype) * scale
+
+    if not op.supports(
+        xformers.ops.AttentionOpDispatch.from_arguments(
+            query=query, key=key, value=value
+        )
+    ):
+        pytest.skip("unsupported configuration")
+
+    _, lse, _, _ = op.FORWARD_OPERATOR(query, key, value, True, None, 0.0)
+    ref_lse = (
+        (query.float() / k_len**0.5) @ key.float().transpose(-2, -1)
+    ).logsumexp(-1)
+
+    assert_allclose(lse[:, : ref_lse.shape[1]], ref_lse, atol=2e-4)
+
+
+@pytest.mark.parametrize(
+    "attn_bias_type", [None, xformers.ops.LowerTriangularMask, torch.Tensor]
+)
+@pytest.mark.parametrize("grad_out_contiguous", [False, True])
+@pytest.mark.parametrize("k_len", [5, 6, 32, 128])
+@pytest.mark.parametrize("batch_size", [1, 4])
+@pytest.mark.parametrize("kv_len", [3, 15, 32, 33, 64, 128])
+@pytest.mark.parametrize("q_len", [2, 3, 5, 32, 128])
+@pytest.mark.parametrize("device", _devices)
+@pytest.mark.parametrize("dtype", [torch.float, torch.half, torch.bfloat16])
+@pytest.mark.parametrize(
+    "op",
+    [
+        xformers.ops.MemoryEfficientAttentionOp,
+        xformers.ops.MemoryEfficientAttentionGenericForwardOp,
+        xformers.ops.MemoryEfficientAttentionFlashAttentionOp,
+    ],
+)
+def test_memory_efficient_attention_backward(
+    device,
+    q_len,
+    kv_len,
+    batch_size,
+    k_len,
+    grad_out_contiguous,
+    attn_bias_type,
+    dtype,
+    op: xformers.ops.MemoryEfficientAttentionOp,
+):
+    torch.manual_seed(batch_size * q_len + kv_len * k_len)
+    scale = 3
+    query = torch.randn((batch_size, q_len, k_len), device=device, dtype=dtype) * scale
+    key = torch.randn((batch_size, kv_len, k_len), device=device, dtype=dtype) * scale
+    value = torch.randn((batch_size, kv_len, k_len), device=device, dtype=dtype) * scale
+
+    attn_bias = None
+    if attn_bias_type is not None:
+        attn_bias = create_attn_bias(
+            attn_bias_type,
+            batch_size=batch_size,
+            q_len=q_len,
+            kv_len=kv_len,
+            dtype=dtype,
+            device=device,
+        )
+
+    if not op.supports(
+        xformers.ops.AttentionOpDispatch.from_arguments(
+            query=query, key=key, value=value, attn_bias=attn_bias
+        )
+    ):
+        pytest.skip("unsupported configuration")
+
+    query.requires_grad_(True)
+    key.requires_grad_(True)
+    value.requires_grad_(True)
+
+    grad_out = torch.ones_like(query)
+    if grad_out_contiguous is False:
+        grad_out = torch.tensor([1.0], device=device)[None, None, :].expand_as(query)
```

Comment:
I must be missing something, but it's not super clear to me that this makes grad_out non contiguous, since it keeps the last dimension as is (unless this comes from the .expand_as() ? it's not clear to me why that would break it)
<img width="511" alt="Screen Shot 2022-08-13 at 11 57 28 PM" src="https://user-images.githubusercontent.com/3072861/184512009-3b7e8a57-ddc9-49f4-a7ea-650ab7f5cbda.png">